src/Pact/Analyze/Check.hs

```diff
@@ -742,46 +799,174 @@ verifyFunctionProps
+    _ ->
       pure $ Right $ TableMap Map.empty
 
 -- TODO: use from Control.Monad.Except when on mtl 2.2.2
 liftEither :: MonadError e m => Either e a -> m a
 liftEither = either throwError return
 
+getConsts
+  :: HM.HashMap Text Ref
+  -> ExceptT VerificationFailure IO (HM.HashMap Text EProp)
+getConsts defconstRefs = do
+
+  (consts :: HM.HashMap Text (AST Node)) <- ifoldrM
+    (\name ref accum -> do
+      maybeConst <- lift $ runTC 0 False $ typecheckTopLevel ref
+      pure $ case maybeConst of
+        (TopConst _info _qualifiedName _type val _doc, _tcState)
+          -> accum & at name ?~ val
+        _ -> error "invariant failure: anything but a const is unexpected here"
+    )
+    HM.empty
+    defconstRefs
+
+  let constToProp :: ETerm -> Except VerificationFailure EProp
+      constToProp tm = case constantToProp tm of
+        Right prop -> pure prop
+        Left msg   -> throwError $ FailedConstTranslation msg
+
+      translateNodeNoGraph'
+        = withExceptT translateToVerificationFailure . translateNodeNoGraph
+
+  hoist generalize $
+    traverseOf each (constToProp <=< translateNodeNoGraph') consts
+
+getStepChecks
+  :: [Table]
+  -> HM.HashMap Text EProp
+  -> HM.HashMap Text (DefinedProperty (Exp Info))
+  -> (CheckableType -> Text -> [Located Check] -> Fun Node -> IO [CheckResult])
+  -> HM.HashMap Text Ref
+  -> ExceptT VerificationFailure IO (HM.HashMap (Text, Int) [CheckResult])
+getStepChecks tables consts propDefs verifyFunProps defpactRefs = do
+
+  (steps :: HM.HashMap (Text, Int)
+    ((AST Node, [Named Node], Info), Pact.FunType TC.UserType))
+    <- ifoldrM
+    (\name ref accum -> do
+      maybeDef <- lift $ runTC 0 False $ typecheckTopLevel ref
+      pure $ case maybeDef of
+        (TopFun (FDefun info _ _ Defpact funType args steps) _meta, _)
+          -> ifoldl
+            (\i stepAccum step ->
+              stepAccum & at (name, i) ?~ ((step, args, info), funType))
+            accum
+            steps
+        _ -> error
+          "invariant failure: anything but a function is unexpected here"
+    )
+    HM.empty
+    defpactRefs
+
+  (stepChecks :: HM.HashMap (Text, Int)
+    ((AST Node, [Named Node], Info), Either ParseFailure [Located Check]))
+    <- hoist generalize $ for steps $ \((step, args, info), pactType) ->
+      case step of
+        TC.Step _ _ exec _ _ model -> ((exec,args,info),) <$>
+          stepCheck tables consts propDefs pactType model
+        _ -> error
+          "invariant violation: anything but a step is unexpected in stepChecks"
+
+  stepChecks' <- case traverse sequence stepChecks of
+    Left errs         -> throwError $ ModuleParseFailure errs
+    Right stepChecks' -> pure stepChecks'
+
+  lift $ ifor stepChecks' $
+    \(name, _stepNum) ((node, args, info), checks) ->
+     verifyFunProps CheckPactStep name checks $
+       -- Only _fInfo, _fArgs, and _fBody are used:
+       FDefun info undefined undefined undefined undefined args [node]
+
+getFunChecks
```

Comment:
Hm -- it seems the only difference is that `getFunChecks` takes one additional argument, the `ModuleData Ref`? Since this combination of args is used twice, maybe we extract a data type and provide that same record in the two calls to these functions? They are both called from the same spot.

Reply:
This is now `CheckEnv`